core/iwasm/common/wasm_c_api.c

```diff
@@ -327,45 +327,80 @@ wasm_engine_new_internal(mem_alloc_type_t type, const MemAllocOption *opts)
 
 /* global engine instance */
 static wasm_engine_t *singleton_engine = NULL;
+#if defined(OS_THREAD_MUTEX_INITIALIZER)
+static korp_mutex engine_lock = OS_THREAD_MUTEX_INITIALIZER;
+#else
+static korp_mutex engine_lock;
+#endif
 
 own wasm_engine_t *
-wasm_engine_new()
+wasm_engine_new_with_args(mem_alloc_type_t type, const MemAllocOption *opts)
 {
+#if !defined(OS_THREAD_MUTEX_INITIALIZER)
+    if (!singleton_engine) {
+        if (os_mutex_init(&engine_lock) != 0) {
+            return NULL;
+        }
+    }
+#endif
+
+    os_mutex_lock(&engine_lock);
+
     if (!singleton_engine) {
-        singleton_engine =
-            wasm_engine_new_internal(Alloc_With_System_Allocator, NULL);
+        singleton_engine = wasm_engine_new_internal(type, opts);
     }
-    if (singleton_engine)
+
+    if (singleton_engine) {
         singleton_engine->ref_count++;
+    }
+
+    os_mutex_unlock(&engine_lock);
+
+#if !defined(OS_THREAD_MUTEX_INITIALIZER)
```

Comment:
It seems there is less trouble if we didn't `os_mutex_destory` `engine_lock` than if we did.

if run `os_mutex_destory`, there are conflicts have to sync:
- multiple `os_mutex_init`. *Attempting to initialize an already initialized mutex results in undefined behavior.*
- multiple 'os_mutex_destory`. *The results of otherwise referencing the object after it has been destroyed are undefined.*
- `os_mutex_destory` in `wasm_engine_delete` and `os_mutex_lock` in `wasm_engine_new_xxx`. 
- and so on.

If don't call `os_mutex_destroy`, beside a logically uncomplete problem(there is no leak), all conflicts should be worried is multiple `os_mutex_init`. 

And for multiple `os_mutex_init`, maybe use `os_mutex_lock` to detect. It will return `EINVAL` for uninitialized object for sure.  Or just leave it there.

Reply:
主要问题是像Windows一样的系统，mutex_init调用CreateMutex申请资源，如果这边没有destroy，就会有资源没有释放的情况。